### PR TITLE
Ks 20191119 scheduler

### PR DESCRIPTION
**Pull request by fil512**
ready for review.

### CODE REVIEW

File: hapi-fhir-base/pom.xml [hunk 1]
```diff
@@ -115,6 +115,11 @@
 			<artifactId>logback-classic</artifactId>
 			<optional>true</optional>
 		</dependency>
+		<!-- Scheduled tasks -->
+		<dependency>
+			<groupId>org.quartz-scheduler</groupId>
+			<artifactId>quartz</artifactId>
```
Comment (jamesagnew) [inline hapi-fhir-base/pom.xml]: Why is this dependency moved to hapi-fhir-base? That means that anyone who wants to use any part of the library (say, a client using just the parser on a mobile device) needs to import quartz and all of its dependencies.

This should stay in one of the jpaserver projects for sure

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/model/api/ISmartLifecyclePhase.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package ca.uhn.fhir.model.api;
+
+public interface ISmartLifecyclePhase {
```
Comment (jamesagnew) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/model/api/ISmartLifecyclePhase.java]: There is no spring stuff outside of JPAServer, this should live in there somewhere

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/sched/BaseHapiScheduler.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			throw new ConfigurationException("Failed to start up scheduler", e);
+		}
+	}
+
+	@Override
+	public void shutdown() {
+		try {
+			myScheduler.shutdown(true);
+		} catch (SchedulerException e) {
+			ourLog.error("Failed to shut down scheduler");
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/sched/BaseHapiScheduler.java]: Should probably log the exception itself here

Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/sched/BaseHapiScheduler.java]: (And above)

Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/sched/BaseHapiScheduler.java]: (and below.. haha)

File: hapi-fhir-base/pom.xml (hunk 1 above)
Comment (fil512) [reply hapi-fhir-base/pom.xml]: done.  (Still need to change cdr)

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/sched/BaseHapiScheduler.java (hunk 3 above)
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/sched/BaseHapiScheduler.java]: good catch.  thanks

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/sched/BaseSchedulerServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+	protected abstract IHapiScheduler getClusteredScheduler();
+
+	/**
+	 * We defer startup of executing started tasks until we're sure we're ready for it
+	 * and the startup is completely done
+	 */
+
+	@Override
+	public int getPhase() {
+		return ISmartLifecyclePhase.SCHEDULER_1000;
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/sched/BaseSchedulerServiceImpl.java]: Did you decide you were going to scrap the smart lifecycle stuff? No worries if not, but if so I guess this should be removed

**Comment by fil512** [comment]
failed on an intermittent but it passed the last two times

**Pull request merged**